core/server/data/migrations/versions/4.0/21-solve-orphaned-webhooks.js

```diff
@@ -0,0 +1,92 @@
+const {createIrreversibleMigration} = require('../../utils');
+const logging = require('../../../../../shared/logging');
+const ObjectID = require('bson-objectid');
+const crypto = require('crypto');
+
+// createSecret is copied from core/server/models/api-key.js
+const createSecret = (type) => {
+    const bytes = type === 'content' ? 13 : 32;
+    return crypto.randomBytes(bytes).toString('hex');
+};
+
+module.exports = createIrreversibleMigration(
+    async function up(knex) {
+        logging.info('Resolving the orphaned webhooks');
+
+        const orphanedWebhooks = await knex('webhooks')
+            .select('webhooks.id')
+            .leftJoin('integrations', 'integrations.id', 'webhooks.integration_id')
+            .where('integrations.id', null);
+
+        if (orphanedWebhooks.length === 0) {
+            logging.info('No orphaned webhooks found, skipping');
+            return;
+        }
+
+        const [role] = await knex('roles')
+            .select('roles.id')
+            .where('name', 'Admin Integration');
+
+        if (!role) {
+            logging.info('Failed to apply migration because of missing role.');
```

Comment:
I would maybe make this `logging.warn/error`? My internal thinking for `info` would be messages where thing are happening as expected. However given the message, I'm not sure which one out of `error`/`warn` it should be. Technically the migration cannot continue due to bad state, so I'd expect error, but we don't want to start logging out errors if there's nothing they can do. @naz What do you think?

Reply:
We don't have a precedent for logging `error` levels within migrations. I think `warn` is the level we usually go with when the state is not what one expects.

Reply:
I've switched to `warn` 👍